better documentation about how external services are integrated (#913)
concord-consortium/rigse: docs/exernal-services.md

# Summary
There are 5 types of external services supported by the Portal:
- runtime environments
- authoring environments
- teacher and student reports or dashboards
- researcher reports
- SSO clients

# Runtime Environments
A runtime environment is a site that handles students running a resource. In some cases the authoring environment and the runtime environment will be the same. The main way the Portal works with a runtime environment is through ExternalActivities. Additionally the Portal has Tools. An ExternalActivity can be associated with one Tool. The Tool provides a way for multiple ExternalActivities to have similar behavior.

Authors create these ExternalActivities (called Resources in the UI).

> Teachers find ExternalActivities and assign them to their class. This assignment is called an Offering in the code. In LTI terminology this is the Offering is the same as a Resource Link.

Reviewer: I think the last sentence should say "In LTI **terminology the Offering** is the same as a Resource Link.", right?

Otherwise, the `ExternalActivity#url` is used. The added parameters can be seen here:
https://github.com/concord-consortium/rigse/blob/d81d5a1da69253418abde7e0aa7baac4a1831a66/rails/app/models/external_activity.rb#L162

## Publishing
The Authoring environment can publish its resources to the Portal. The publish creates or updates an ExternalActivity.

## Reporting
There are two ways reports can be associated with the ExternalActivity:
- directly referencing an ExternalReport
- having a Tool with a `source_type` that matches the `default_report_for_source_type` of an ExternalReport

More details on reports are below.

## Authorization
The runtime environment will likely need to use Portal APIs. It might need them to find out who the members of the class are. Or it might want the portal to generate a JWT the runtime can use with Firebase.

### Non LARA Runtime
For a non LARA launch this is usually achieved by passing a short lived token. Then the runtime can send this token back to the portal in a Authorization header. If the runtime needs to access more APIs after the initial launch it can request a Portal JWT that lasts longer and can be renewed.

> The token is generated by making a AccessGrant. And the token is a property of the grant. In this case the AccessGrant doesn't have a client. The API Controller's authentication check does *not* verify the client of a token. However the general Bearer token authentication of the Portal *does* check for a client. So this short lived token can only be used with endpoints that use the API Controller's authentication check. Not all controllers in `api/vi` use the API Controllers auth check.  The best approach is to exchange the short lived token for a Portal JWT which can then work with all apis.

Reviewer: "The token is generated by making a**n** AccessGrant"

References (search for these strings)
- token generation: `create_access_token_with_learner_valid_for`
- api controller auth check: `check_for_auth_token`
- general token check: `token_valid?`

Example [implementation in CLUE](https://github.com/concord-consortium/collaborative-learning/blob/14d066421c6f6a55102bb9d1a9d849491928316e/src/lib/auth.ts#L310)

### LARA Runtime

> For a LARA launch, OAuth2 is used. The portal sends a domain and domain_uid parameter to LARA. LARA checks if there is a user currently signed in with a auth client for this domain and this uid. If not then LARA redirects the user back to the Portal to get signed in. And then the user is returned back to LARA with a AccessGrant token. The URL parameters are preserved though out this process. In this case the AccessGrant has a client, so LARA can use this token with any API.

Reviewer 2: ```suggestion
For a LARA launch, OAuth2 is used. The portal sends a `domain` and `domain_uid` parameter to LARA. LARA checks if there is a user currently signed in with an auth client for this domain and this uid. If not then LARA redirects the user back to the Portal to get signed in. And then the user is returned back to LARA with an AccessGrant token. The URL parameters are preserved throughout this process. In this case the AccessGrant has a client, so LARA can use this token with any API.
```

> For a LARA launch, OAuth2 is used. The portal sends a domain and domain_uid parameter to LARA. LARA checks if there is a user currently signed in with a auth client for this domain and this uid. If not then LARA redirects the user back to the Portal to get signed in. And then the user is returned back to LARA with a AccessGrant token. The URL parameters are preserved though out this process. In this case the AccessGrant has a client, so LARA can use this token with any API.

Reviewer: "...back to LARA with a**n** AccessGrant"

The LARA runtime javascript code does not use this token. Instead it makes requests to the LARA server and the LARA server then makes a request to the Portal.

# Authoring environments
The only authoring environment the portal currently has is LARA. However other environments should almost work. The main integration point for these environments is the `ExternalActivity#author_url`. If this is set then the Portal will open this URL when an author clicks the edit button or link associated with the external activity.

## Publishing
Authoring environments can also 'publish' resources to the portal. Currently LARA is the only system doing this. LARA uses 2 methods of publishing.

### LARA Runtime
For the LARA built in runtime it uses the portal publish api. This not only includes information for setting up the ExternalActivity, but also includes the reportable structure of the resource. The Portal takes this structure and creates or updates a set of models to save the structure.

### LARA Activity Player Runtime

> For the Activity Player runtime, LARA uses the more basic ExternalActivity api. In this case it simply creates or updates the external activity.  This is same api that should be used by non LARA systems that want to managed ExternalActivities to make it easier for authors.

Reviewer 2: ```suggestion
For the Activity Player runtime, LARA uses the more basic ExternalActivity api. In this case it simply creates or updates the external activity.  This is the same api that should be used by non LARA systems that want to manage ExternalActivities to make it easier for authors.
```

## AuthoringSite model

> Additionally there is an AuthoringSite model is how a convenience 'create XYZ' button can be added to `/authoring` page.

Reviewer 2: ```suggestion
Additionally there is an AuthoringSite model, which is how a convenience 'create XYZ' button can be added to `/authoring` page.
```

# Teacher and Student Reports or Dashboards
As described above ExternalReports are used to provide reports and dashboards for ExternalActivities. The ExternalReport model is also used for Researcher reports.

The ExternalReport has a Client which indicates what domain the report is hosted on, so the portal can validate its requests.

## Types
An ExternalReport has 2 types for teachers:
- offering
- class

It can also be marked as "allowed for students"

The other types of ExternalReports are described in the Researcher section

### Offering Report

> An offering report is associated with an assignment in the Portal. So in places where teachers see the assignment they also see a button to run this report.

Reviewer 2: ```suggestion
An offering report is associated with an assignment in the Portal. In places where teachers see the assignment they also see a button to run this report.
```

The report has to be associated with the assigned ExternalActivity either directly or indirectly through the source_type.

### Class Report
A class report is for classes. In places where teachers see classes they will see a button to run this report.

> The report has to be associated directly with the ExternalActivity. The indirect use of source_type is not supported for Class reports.

Reviewer 2: ```suggestion
The report has to be associated directly with the ExternalActivity. The indirect use of `source_type` is not supported for Class reports.
```

> When a class has 1 or more assignments with a Class Report then the report button is shown at the class level of the UI.

Reviewer 2: ```suggestion
When a class has one or more assignments with a Class Report, then the report button is shown at the class level of the UI.
```

### Student Report
If the report is allowed for students, and the report is an offering report, then the student will see a report button next to the assignment.

When the student report is launched, the user id of the current user is passed as a `studentId` to the report. This is in addition to the other parameters below. Reference: https://github.com/concord-consortium/rigse/blob/d81d5a1da69253418abde7e0aa7baac4a1831a66/rails/app/models/external_report.rb#L39
Also the learner (student, offering) is added to the access grant of the token.

## Launching
The offering_controller provides the following actions for reports:
- student_report
- report
- external_report

The Portal adds several parameters onto the URL provided by the ExternalReport. They can be seen here: https://github.com/concord-consortium/rigse/blob/d81d5a1da69253418abde7e0aa7baac4a1831a66/rails/app/models/external_report.rb#L68

## Authorization

> Like ExternalActivities, ExternalReports provide a token the report can use to call back to the Portal. Unlike ExternalActivities this token has a client. So when an ExternalReport is setup in the Portal a Client is needed and must have the correct settings. Our teacher reports are normally SPAs, so the Client should have a type of public, and it should specify the domain of the SPA.

Reviewer 2: ```suggestion
Like ExternalActivities, ExternalReports provide a token the report can use to call back to the Portal. Unlike ExternalActivities this token has a client. So when an ExternalReport is set up in the Portal a Client is needed and must have the correct settings. Our teacher reports are normally SPAs, so the Client should have a type of `public`, and it should specify the domain of the SPA.
```

> If the SPA report is also going to support OAuth2, so it can be run without being launched from the portal with a token, then the redirect url must be specified. This redirect url is used by the Portal after the user is signed in.

Reviewer 2: ```suggestion
If the SPA report is also going to support OAuth2 (so it can be run without being launched from the portal with a token), then the redirect url must be specified. This redirect url is used by the Portal after the user is signed in.
```

Additionally if the user is a teacher then the teacher object is added to the access grant.  If the user is a student then the learner object (student, offering) is added to the access grant. These additions to the grant are used when the report requests a JWT. The issued JWT then includes claims based on the teacher or learner. References:
- https://github.com/concord-consortium/rigse/blob/d81d5a1da69253418abde7e0aa7baac4a1831a66/rails/app/controllers/api/v1/jwt_controller.rb#L59
- https://github.com/concord-consortium/rigse/blob/d81d5a1da69253418abde7e0aa7baac4a1831a66/rails/app/controllers/api/v1/jwt_controller.rb#L89

# Researcher Reports

**TODO**

# SSO clients

**TODO**

# From a Modeling Point of View

## Tool

> In the portal there is a tools model. This model has a tool_id (url), source_type, and name. An ExternalActivity can be associated with a single Tool.

Reviewer 2: ```suggestion
In the portal there is a `tools` model. This model has a `tool_id` (url), `source_type`, and `name`. An ExternalActivity can be associated with a single Tool.
```

The source_type of the ExternalActivity#tool is used to find an ExternalReport configured with a default_report_for_source_type.

> In the user researcher report filer page, only ExternalActivites that have a Tool with a source_type of "LARA" are shown in the "runnables" list.

Reviewer 2: ```suggestion
In the user researcher report filer page, only ExternalActivites that have a Tool with a `source_type` of "LARA" are shown in the "runnables" list.
```

> When the (current defunct) move student feature is used, the portal sends the tool_id (url) of the ExternalActivities in the original class and new class of the student, to the ExternalReports#move_students_api_url

Reviewer 2: ```suggestion
When the (currently defunct) move student feature is used, the portal sends the `tool_id` (url) of the ExternalActivities in the original class and new class of the student, to the ExternalReports#move_students_api_url
```

> When the (current defunct) move student feature is used, the portal sends the tool_id (url) of the ExternalActivities in the original class and new class of the student, to the ExternalReports#move_students_api_url

Reviewer: "(current**ly** defunct)"

> The tool_id is used when an external activity is created using the `api/v1/external_activity` api.

Reviewer 2: ```suggestion
The `tool_id` is used when an external activity is created using the `api/v1/external_activity` api.
```

> When ActivityRuntimeAPI is used to publish a new activity or sequence a new Tool is created if the source_type is passed in. In this case the Tool has no tool_id. This feature is mainly for developer convenience.

Reviewer 2: ```suggestion
When ActivityRuntimeAPI is used to publish a new activity or sequence a new Tool is created if the `source_type` is passed in. In this case the Tool has no `tool_id`. This feature is mainly for developer convenience.
```

## ExternalReport

> Has a `default_report_for_source_type`. When an ExternalActivity has a tool with this source_type is assigned to a class, then report buttons for this assignment will automatically be added.

Reviewer 2: ```suggestion
Has a `default_report_for_source_type`. When an ExternalActivity has a tool with this `source_type` is assigned to a class, then report buttons for this assignment will automatically be added.
```

> ExternalActivities can also have 1 or more ExternalReports directly associated with them. This makes it possible for an author or admin to setup a special report or dashboard for a specific resource.

Reviewer 2: ```suggestion
ExternalActivities can also have one or more ExternalReports directly associated with them. This makes it possible for an author or admin to set up a special report or dashboard for a specific resource.
```

## ExternalActivity

> Authors and admins can duplicate ExternalActivities. If the ExternalActivity#tool.source_type is "LARA" then the a special `duplicate_on_lara` function is called. This sends a http request to `#{ExternalActivity#url}/remote_duplicate`. It expects LARA to copy the sequence or activity and then return the same kind of data that would normally be sent when an author "publishes" the sequence or activity to the Portal. Then the Portal does this publishing internally using the `ActivityRuntimeAPI`. This internal publishing happens as a 'update' operation not a 'create' operation. This is because the ExternalActivity already exists.

Reviewer 2: ```suggestion
Authors and admins can duplicate ExternalActivities. If the ExternalActivity#tool.source_type is "LARA" then the special `duplicate_on_lara` function is called. This sends a http request to `#{ExternalActivity#url}/remote_duplicate`. It expects LARA to copy the sequence or activity and then return the same kind of data that would normally be sent when an author "publishes" the sequence or activity to the Portal. Then the Portal does this publishing internally using the `ActivityRuntimeAPI`. This internal publishing happens as an 'update' operation not a 'create' operation. This is because the ExternalActivity already exists.
```

> Authors and admins can duplicate ExternalActivities. If the ExternalActivity#tool.source_type is "LARA" then the a special `duplicate_on_lara` function is called. This sends a http request to `#{ExternalActivity#url}/remote_duplicate`. It expects LARA to copy the sequence or activity and then return the same kind of data that would normally be sent when an author "publishes" the sequence or activity to the Portal. Then the Portal does this publishing internally using the `ActivityRuntimeAPI`. This internal publishing happens as a 'update' operation not a 'create' operation. This is because the ExternalActivity already exists.

Reviewer: "publishing happens as a**n** 'update' operation"

> When the ActivityRuntimeAPI is used to create or update an ExternalActivity, the `launch_url` is only set when it is created. Because of this the duplicate_on_lara code sets the `launch_url` first before doing the internal publish.

Reviewer 2: ```suggestion
When the ActivityRuntimeAPI is used to create or update an ExternalActivity, the `launch_url` is only set when it is created. Because of this the `duplicate_on_lara` code sets the `launch_url` first before doing the internal publish.
```

> When data_helpers.rb generates JSON from the activity it includes lara_activity_or_sequence which is based on whether the activity has `launch_url`. This seems be used to determine which links are shown to the user.

Reviewer 2: ```suggestion
When data_helpers.rb generates JSON from the activity it includes `lara_activity_or_sequence` which is based on whether the activity has `launch_url`. This seems be used to determine which links are shown to the user.
```

## Client

> Clients are used for authentication. When the Portal creates an AccessGrant to provide a token to an external service, it generally has an associated Client. The Client is used to verify requests with the token. If the client specifies domain_matchers then only requests from those domains are accepted.

Reviewer 2: ```suggestion
Clients are used for authentication. When the Portal creates an AccessGrant to provide a token to an external service, it generally has an associated Client. The Client is used to verify requests with the token. If the client specifies `domain_matchers` then only requests from those domains are accepted.
```

When the request is part of an OAuth2 authorize, then the Client needs to specify the exact redirect that the Portal is sending this token or code back to. These OAuth2 authorize requests are used when LARA is doing OAuth2. They are also used by the single page apps that use the Portal for authentication like the token-service, glossary authoring, or a portal-report when not launched directly from the portal.

## Seeding the Database
To make it easier for developers, when the database is seeded a few of the resources above are automatically created.

An ExternalReport is created for the portal-report. And a Client is created that this ExternalReport uses.

# Ways to Improve

## Eliminate the source_type

### default_report

> The use of source_type to connect ExternalActivities to default reports is not needed. Instead a 'default_report' can be added to the Tool model. This will have the same behavior and simplify the modeling.

Reviewer 2: ```suggestion
The use of `source_type` to connect ExternalActivities to default reports is not needed. Instead a `default_report` can be added to the Tool model. This will have the same behavior and simplify the modeling.
```

### remove filtering of runnables in user researcher report

> Currently the runnables are filtered by source_type in the user researcher report page. This is not really useful, and seems could just be removed. If we need filtering like this, we could add Tool filtering

Reviewer 2: ```suggestion
Currently the runnables are filtered by `source_type` in the user researcher report page. This is not really useful, and seems like it could just be removed. If we need filtering like this, we could add Tool filtering.
```

### duplicate activity support

> Currently a source_type of LARA means the portal uses the `remote_duplicate` action in LARA when an ExternalActivity is duplicated. It is convenient for authors so they can duplicate resources in the portal without need to setup the ExternalActivity for the resource (description, grade_levels, ...). Instead of using this check on a source_type of LARA, this should be made generic so non-LARA authoring systems can provide support for duplicating from the Portal.

Reviewer 2: ```suggestion
Currently a `source_type` of LARA means the portal uses the `remote_duplicate` action in LARA when an ExternalActivity is duplicated. It is convenient for authors so they can duplicate resources in the portal without needing to set up the ExternalActivity for the resource (description, grade_levels, ...). Instead of using this check on a `source_type` of LARA, this should be made generic so non-LARA authoring systems can provide support for duplicating from the Portal.
```
